src/utils/daemon.js

```diff
@@ -1,63 +1,104 @@
-  opts.type = opts.type || 'go'
-  opts.path = opts.path || ''
-  opts.flags = opts.flags || ['--migrate=true', '--routing=dhtclient']
-  opts.keysize = opts.keysize || 4096
+function repoFsck (path) {
+  const exec = findExecutable('go', join(__dirname, '..'))
+  spawnSync(exec, ['repo', 'fsck'], {
+    env: {
+      ...process.env,
+      IPFS_PATH: path
+    }
+  })
+}
 
-  if (opts.type !== 'go') {
-    throw new Error(`${opts.type} connection is not supported yet`)
+async function configure (ipfsd) {
+  const cfgFile = join(ipfsd.repoPath, 'config')
+  const cfg = await fs.readJSON(cfgFile)
+
+  let origins = []
+  try {
+    origins = cfg.API.HTTPHeaders['Access-Control-Allow-Origin']
+  } catch (e) {
+    logger.warn(e)
   }
 
-  const factory = IPFSFactory.create({ type: opts.type })
+  if (!Array.isArray(origins)) {
+    origins = []
+  }
+
+  if (!origins.includes('webui://-')) origins.push('webui://-')
+  if (!origins.includes('https://webui.ipfs.io')) origins.push('https://webui.ipfs.io')
+
+  cfg.API.HTTPHeaders['Access-Control-Allow-Origin'] = origins
+  cfg.API.HTTPHeaders['Access-Control-Allow-Methods'] = ['PUT', 'GET', 'POST']
+
+  await fs.writeJSON(cfgFile, cfg)
+}
+
+async function spawn ({ type, path, keysize = 0 }) {
+  const factory = IPFSFactory.create({ type: type })
 
-  const ipfsd = await new Promise((resolve, reject) => {
+  return new Promise((resolve, reject) => {
     factory.spawn({
       disposable: false,
       defaultAddrs: true,
-      repoPath: opts.path
+      repoPath: path
     }, (e, ipfsd) => {
       if (e) return reject(e)
       if (ipfsd.initialized) {
         return resolve(ipfsd)
       }
 
       ipfsd.init({
-        directory: opts.path,
-        keysize: opts.keysize
+        directory: path,
+        keysize: keysize
       }, e => {
         if (e) return reject(e)
         resolve(ipfsd)
       })
     })
   })
+}
 
-  if (!ipfsd.started) {
-    await new Promise((resolve, reject) => {
-      ipfsd.start(opts.flags, err => {
-        if (err) {
-          return reject(err)
-        }
+async function start (ipfsd, { flags }) {
+  await new Promise((resolve, reject) => {
+    ipfsd.start(flags, err => {
+      if (err) {
+        return reject(err)
+      }
 
-        resolve()
-      })
+      resolve()
     })
+  })
+}
+
+export default async function (opts) {
+  const ipfsd = await spawn(opts)
+  await configure(ipfsd)
+
+  if (!ipfsd.started) {
+    await start(ipfsd, opts)
   }
 
-  let origins = []
   try {
-    origins = await ipfsd.api.config.get('API.HTTPHeaders.Access-Control-Allow-Origin')
+    await ipfsd.api.id()
   } catch (e) {
-    logger.warn(e)
-  }
+    if (!e.message.includes('ECONNREFUSED')) {
+      throw e
+    }
 
-  if (!origins.includes('webui://-')) origins.push('webui://-')
-  if (!origins.includes('https://webui.ipfs.io')) origins.push('https://webui.ipfs.io')
+    if (!showRepoApiFileErrorMessage(ipfsd.repoPath, ipfsd.apiAddr)) {
```

Comment:
This is neat, but it's now assuming that there is an `api` file rather than checking for it. Either we need to check for it so we can assert it in the error message, or we need to update the wording of the error message to something more like:

"IPFS Desktop failed to connect to an existing ipfs api at ${ipfsd.apiAddr}. This can happen if you run `ipfs daemon` manually and it is not shutdown cleanly. Would you like to try running `ipfs repo fsck` to remove the lock and api files from ${ipfsd.repoPath} and try again?

Reply:
Just updated the error message.